snippets/titration/titration.py

```diff
@@ -0,0 +1,947 @@
+    mask_list.append(tifffile.imread(mask_paths[i]))
+    
+    if image_list[i].shape[0] != len(ad.var_names):
+        print("The panel and image have a different number of channels")
+        
+    # Subset the channel corresponding to the curent marker
+    image_list[i] = image_list[i][channel]
+    
+    if image_list[i].shape != mask_list[i].shape:
+        print("The mask and the image have different shapes")
+    
+# Crop the images
+crop_mask = []
+for i,conc in enumerate(concentrations):
+    if(shape(image_list[i])[1] > crop_width):
+        rdm_x1 = np.random.randint(0, shape(image_list[i])[1] - crop_width)
+        rdm_x2 = rdm_x1 + crop_width
+    else:
+        rdm_x1 = 0
+        rdm_x2 = shape(image_list[i])[1]
+        
+    if(shape(image_list[i])[0] > crop_height):
+        rdm_y1 = np.random.randint(0, shape(image_list[i])[0] - crop_height)
+        rdm_y2 = rdm_y1 + crop_height
+    else:
+        rdm_y1 = 0
+        rdm_y2 = shape(image_list[i])[0]
+    
+    crop_mask.append([rdm_y1, rdm_y2, rdm_x1, rdm_x2])
+
+# %% [markdown]
+# ###  Mask cells
+# Retrieve the positive and negative cells and create a mask overlay.
+
+# %%
+method_column = w_method.value
+
+positive_cells = numpy.empty(len(concentrations)).tolist()
+intermediate_cells = numpy.empty(len(concentrations)).tolist()
+negative_cells = numpy.empty(len(concentrations)).tolist()
+mask_overlays = [0] * len(concentrations)
+
+for i, conc in enumerate(concentrations):
+    # Retrieve the ids of positive and negative cells
+    ad_subset = ad[(ad.obs['concentration'] == conc) & (ad.obs['image_id'] == selected_images[i]),
+                   ad.var_names == cur_marker]
+    positive_cells[i] = ad_subset[ad_subset.obs[method_column] == 'Positive'].obs['object_id']
+    intermediate_cells[i] = ad_subset[ad_subset.obs[method_column] == 'Inter'].obs['object_id']
+    negative_cells[i] = ad_subset[ad_subset.obs[method_column] == 'Negative'].obs['object_id']
+    
+    # Mask the cells
+    pos = np.array(positive_cells[i]).astype('int')
+    inter = np.array(intermediate_cells[i]).astype('int')
+    neg = np.array(negative_cells[i]).astype('int')
+
+    mask_overlays[i] = np.empty(shape(mask_list[i]), dtype=int)
+    mask_overlays[i].fill(0)
+    mask_overlays[i][np.isin(mask_list[i], pos)] = 3
+    mask_overlays[i][np.isin(mask_list[i], inter)] = 2
+    mask_overlays[i][np.isin(mask_list[i], neg)] = 1
+
+# %% [markdown]
+# ### Display images and masks
+#
+# __Row 1:__ Selected images, the current marker is shown.
+#
+# __Row 2:__ Masks plotted onto images.
+# Positive cells are overlaid in blue, negative cells in red and intermediate cells in yellow.
+#
+# ___Manually adjust image brightness and mask transparency if needed___
+
+# %%
+# Brightness and transparency
+brightness = 20
+transparency = 0.3
+
+# Display images
+fig, axs = plt.subplots(2, len(concentrations), figsize=(25, 25))
+
+for i, conc in enumerate(concentrations):
```

Comment:
You could also do:
```python
for ax, conc in zip(axs.flat, concentrations):
```

Reply:
It somehow doesn't work, maybe I missed something.